README.md
[![Downloads](https://pepy.tech/badge/AxionVueOpenAPI)](https://pepy.tech/project/AxionVueOpenAPI)
[![Downloads](https://pepy.tech/badge/CytoSmartOpenAPI)](https://pepy.tech/project/CytoSmartOpenAPI)
[![Downloads](https://pepy.tech/badge/luxconnector)](https://pepy.tech/project/luxconnector)

# Axion Vue Open API

This is a Python wrapper for the Axion Vue Windows app to use it headless.
The package will only work on Windows 10 and above.

This package was formally known as `luxconnector` or `CytoSmartOpenAPI`

## Warranty

> ⚠️ **Hardware warranty is void by using this open API** ⚠️:
>
> Using the AxionVueOpenAPI means you will have NO hardware warranty (see license).

This is because our devices are made to handle the normal usage ([the GUI app](https://download.axionbio.com/)).
It also includes additionally bought warranty.
Only if your additionally bought warranty implicitly includes the Open API usage you will have warranty.

## Installation

To install this package follow these steps:

### Step 1: Drivers

Make sure you had a [Axion Vue application installed](http://download.axionbio.com/).

Recommended is the cell counter given it does not restart itself after closing.
This is needed to have all the correct drivers installed.
Uninstall the app afterwards.
If the app is on the same machine the Open API might connect with the wrong one.

### Step 2: pip install

```cmd
pip install AxionVueOpenAPI
```

## Initialization

Make sure at least one Axion imaging device is physically connected to your computer via a USB3 port.
The AxionVueOpenAPI is an object that maintains the connection with the device.
To create the object use the following code:

```python
from AxionVueOpenAPI import AxionVueOpenAPI
connector = AxionVueOpenAPI(number_of_devices=2, warranty=False)
```

> ⚠️ You are voiding your warranty by using this package

NOTE: Make sure number_of_devices does not exceed the number of devices you have physically connected.
Otherwise the AxionVueOpenAPI will look for the missing devices forever.

## Getting serial numbers

Each physical device has a serial number.
This number is needed to send commands to a specific device.

```python
serial_numbers = connector.get_all_serial_numbers()
```

## Getting a single image

When you want a single image taken at this moment use get_image.
This will return the image as a [`pillow image`](https://pillow.readthedocs.io/en/stable/reference/Image.html).
You need to give the serial number of the device you want to target.

```python
img = connector.get_image(serial_number)
```

or

```python
img = connector.get_image(serial_numbers[index])
```

## Changing the focus

This function will change the focus for the AxionVueOpenAPI object.
Every image taken after this function will have the new focus.

The focus is in the range 0 until 1. (0.0 and 1.0 are valid entries)

You need to give the serial number of the device you want to target.

```python
connector.set_focus(serial_number, 0.5)
img1 = connector.get_image(serial_number) # Image with focus of 0.5
connector.set_focus(serial_number, 0.7)
img2 = connector.get_image(serial_number) # Image with focus of 0.7
img3 = connector.get_image(serial_number) # Image with focus of 0.7
```

## Auto focus

The device goes over multiple possible focuses to find the best focus.
It will use image analysis to determine how well the image is in focus.

```python
# Fast but goes over a limited range to cover most manufactured cell counting slides
connector.do_autofocus(serial_number, "slide")
# Fastest: goes over the limited range where cells can be in focus on the Axion Exact slide
connector.do_autofocus(serial_number, "CSslide")
# Slow, but goes over the full range of possible focusses: suitable for any vessel that works with the Axion imaging device
connector.do_autofocus(serial_number, "other")
```

## Getting the temperature

This function returns the temperature in Celsius of the device.

You need to give the serial number of the device you want to target.

```python
temperature = connector.get_temperature(serial_number)
```

## Change active camera (fluorescence)

To use fluorescence you will need to change the active camera.
The camera can be set to 3 different values, RED, GREEN, or BRIGHTFIELD.
This will only work if your device is an FL device otherwise only BRIGHTFIELD is available.

```python
connector.set_active_camera(serial_number, "RED")
connector.set_active_camera(serial_number, "GREEN")
connector.set_active_camera(serial_number, "BRIGHTFIELD")
```

## Change camera setting

Each camera has its own settings.
Not all settings are available for BRIGHTFIELD.

- exposure: The time in milliseconds that the camera is detecting light. (Lux and Exact devices only; for Omni devices use flash duration)
- gain: The multiplication of the camera. (Fluo only)
  If very little light goes into the camera sensor make sure the gain is high.
- brightness: Strength of the led when it is on. (Fluo only)
- focus_offset: The difference in focus between brightfield and fluo.
  If focus is set to 0.4 and focus_offset for RED is set to 0.1 RED focus is 0.5 (Fluo only)

```python
connector.set_camera_settings(serial_number, "RED", 500, gain=30, brightness=5000, focus_offset=0.0)
connector.set_camera_settings(serial_number, "BRIGHTFIELD", 10)
```

## Change flash duration (Omni only)

The Omni cannot change the exposure time but can change the time the led is on.
This can be done with set_flash_duration.
It will set the duration between 40 and 250 μs.

```python
connector.set_flash_duration(serial_number, duration=120)
```

If you do NOT have an Omni use set_camera_settings.

## Getting a z-stack

This function will return a list of [`pillow images`](https://pillow.readthedocs.io/en/stable/reference/Image.html).
Each image will be at a different focus level.

This code will create a z-stack of 6 images.
The focuses of these images will be [0.5, 0.6, 0.7, 0.8, 0.9, 1.0]

You need to give the serial number of the device you want to target.

```python
list_of_imgs = connector.get_z_stack(serial_number, num_img = 6, start_focus = 0.5, stop_focus = 1)
```

## Change stage (Omni only)

To change the position of the Omni stage (camera, led, arm, ect.) give the new position in mm.

- Position x in mm
- Position y in mm
- Time in seconds it can wait till it arrives (default 60)

```python
connector.move_stage(serial_number, 100, 100, 60)
```

After this the normal process for obtaining and changing cameras applies.

## Get stage position (Omni only)

If you need to know where the stage is, use get_position.
This also works if the Omni moved without a move_stage command (e.g. when it goes to sleep).

```python
print(connector.get_position(serial_number))
```

## Changing zoom modes (Lux only)

There are 2 zoom modes: "IN" and "OUT".
While zoomed in the resolution is higher but the ROI is smaller, zoomed out has a higher ROI but a lower resolution.
Since zooming in comprises digital zoom, the image will not show more sample details.

Changing this setting will change it for every image or z-stack taken afterwards.

You need to give the serial number of the device you want to target.

```python
connector.set_zoom(serial_number, "IN")
img1 = connector.get_image(serial_number) # Image is zoomed in
connector.set_zoom(serial_number, "OUT")
img2 = connector.get_image(serial_number) # Image is zoomed out
```

## Live view

The live view of the device is hosted at http://localhost:3333/cytosmartservice/live?serialNumber=##########.
This image can only been seen if the live view is turned on (by default the live view is turned on).

You need to give the serial number of the device you want to target at the place of the #-symbols.

```python
connector.set_liveview(serial_number, False) # Led of device turns off till you take a picture
connector.set_liveview(serial_number, True) # In the browser you can see the image being updated
connector.open_liveview(serial_number) # Opens liveview in the default browser
```

## Developers

Developers of the AxionVueOpenAPI please look at the [`developers readme`](README_DEV.md)

## Credits

- Tom "RoadRunner" Nijhof
- Nora
- Tessa
- Marc "it is with a c" van Vijven
- Julia "I did not break it, I tested it" van den Beemd
- Count Nicolai Vondracek, Señior sunshine manager


HISTORY.rst
History
-------
1.3.1 (2023-03-29)
------------------
- Renaming to AxionVueOpenAPI

1.3.0 (2022-08-25)
------------------
- Add open_liveview
- Add do_autofocus
- Add listener to autofocus change
- Bugfix: set_focus. Make sure camera_setting is completed before setting the focus

1.2.0 (2022-06-27)
------------------
- Update server to support Omni
- Add move stage
- Add get position
- Add set flash duration
- Add mandatory turning off of warranty
- Check for connection loss and reconnect is closed
- Check for sleeping device

1.1.1 (2022-06-29)
------------------
- Update documentation

1.1.0 (2021-12-03)
------------------
- Add change camera settings
- Add set activate camera (brightfield, red-fluo, or green-fluo)
- Change focus: 
    make sure focus always happens in brightfield camera mode
    Add sleep after changing to focus so camera can move to new position
- Update to new Cytosmart app

1.0.4 (2021-04-15)
------------------
- Update Pillow
- Fix installation requirements

1.0.3 (2021-11-22)
------------------
- Update requirements of websocket and websocket-client to include version 1
- Check if app is open before starting a new one

1.0.2 (2021-04-15)
------------------
- Update pillow
- Update license to ACADEMIC PUBLIC LICENSE

1.0.1 (2021-01-14)
------------------
- Added support for LUX3 FL
- Bugfix: toggle liveview when change zoom level

1.0.0 (2020-09-15)
------------------
- Added multi lux support
- Changed output from numpy array to pillow
- Require serial number with function to get information
- Add get_temperature

0.1.2 (2020-06-30)
------------------
- Rename failing to pinging in print statement

0.1.1 (2020-06-11)
------------------
- Update documentation
- Make setup.py fit for PyPI
 
0.1.0 (2020-05-20)
------------------
- Update lux server to latest version

0.0.3 (2020-03-30)
------------------
- bugfix: files were not copied into package on linux

0.0.2 (2020-03-24)
------------------
- Bugfix: stop waiting for response after sending a message

0.0.1 (2020-03-16)
------------------

- Add function: Copy the luxconnector 
- Add function: focus change option
- Add function: z-stack
- Bugfix: add lux app to manifest

0.0.0 (2020-03-13)
------------------

Copied basic package a start of luxconnector


AxionVueOpenAPI/main.py
#%%
import json
import os
import subprocess
from io import BytesIO
from pathlib import Path
import time
from typing import List, Tuple
import webbrowser

import requests
from PIL import Image
from websocket import create_connection

from .listener import Listener


class AxionVueOpenAPI:
    def __init__(self, number_of_devices: int = 1, *, warranty: bool) -> None:
        """
        number_of_devices: (int) How many devices should be connected.
            It will keep trying to connect till it is connected to all connected devices.
        warranty: (bool) ⚠️ When using the openAPI you are voiding the hardware warranty
        """
        if warranty:
            raise ValueError("The warranty MUST be set to False")
        self.__connect_with_service()
        self.ws_listener = Listener(self.__recv_ws_message)
        self.ws_listener.start()
        self.__all_devices = self.ws_listener.all_devices
        assert (
            number_of_devices >= 1
        ), f"number_of_devices needs to be above 1, not {number_of_devices}"
        print(f"Connecting to {number_of_devices} devices")

        while True:
            try:
                connected_devices = 0
                all_serial_numbers = self.get_all_serial_numbers()
                for serial_number in all_serial_numbers:
                    img = self.get_image(serial_number)
                    if img is not None:
                        connected_devices += 1

                if connected_devices >= number_of_devices:
                    break

            except:
                pass

    def __connect_with_service(self):
        """
        Try to connect with Axion Vue websocket instance.
        If that does not work, start a instants of the service
        """
        try:
            # Try to make a connection with the app
            self.ws = create_connection("ws://localhost:3333/cytosmartservice")
        except:
            # If that fails, start the app first and make the connection again
            self.__start_axion_vue_app()
            self.ws = create_connection("ws://localhost:3333/cytosmartservice")

    def __send_ws_message(self, msg: dict, count: int = 0):
        """Send a message to the service.
        It will retry if the connection got lost after restoring the connection

        Args:
            msg (dict): Message itself needs to have type and payload
            count (int, optional): Number of times trying. Defaults to 0.

        Raises:
            RecursionError: If a 100 times is not enough to send the message, give up
        """
        if count == 100:
            raise RecursionError(f"Retried to send a message {count} times")
        try:
            self.ws.send(json.dumps(msg))
        except ConnectionResetError:
            print(f"Connection lost, reconnecting. Attempt {count}")
            self.__connect_with_service()
            self.__send_ws_message(msg, count + 1)

    def __recv_ws_message(self, count: int = 0):
        """Receive a message to the service.
        It will retry if the connection got lost after restoring the connection

        Args:
            count (int, optional): Number of times trying. Defaults to 0.

        Raises:
            RecursionError: If a 100 times is not enough to recieve the message, give up
        """
        if count == 100:
            raise RecursionError(f"Retried to receive a message {count} times")

        try:
            return json.loads(self.ws.recv())
        except ConnectionResetError:
            print(f"Connection lost, reconnecting. Attempt {count}")
            self.__connect_with_service()
            self.__recv_ws_message(count + 1)

    def __activate(
        self,
        serial_number: str,
    ) -> None:
        """
        Activate device

        serial_number: (str) the serial number of device you want to connect

        """
        self.__send_ws_message(
            {"type": "ACTIVATE", "payload": {"serialNumber": serial_number}}
        )

    @staticmethod
    def __start_axion_vue_app() -> None:
        """
        Run the Axion Vue server in a subprocess
        """
        print("Start Axion Vue Server")
        basefolder_loc = Path(__file__).parents[0]
        exe_loc = os.path.join(basefolder_loc, "AxionVueApp", "CytoSmartService.exe")
        subprocess.Popen(["cmd", "/K", exe_loc])

    def do_autofocus(
        self, serial_number: str, chamber_type: str, max_waiting_time: float = 60
    ) -> None:
        """
        Turn the liveview on or off

        serial_number: (str) the serial number of device you want to connect
        chamber_type: (str)
            other (SLOW) -> the device will go through its whole range of motion.
            csslide(FAST) -> the device will go through limited range of motion where the cells are expected to be.
            slide (FAST) -> the device will go through limited range of motion where the cells are expected to be.
        max_waiting_time (float, optional): Maximum time it waits for the stage to arrive.
            If it is set to -1 it will never timeout
            Defaults to 60.
        """
        if chamber_type.lower() == "other":
            chamber_type_number = 0
        elif chamber_type.lower() == "csslide":
            chamber_type_number = 1
        elif chamber_type.lower() == "slide":
            chamber_type_number = 2
        else:
            raise ValueError(f"Chamber type {chamber_type} is not supported")

        # Get device and set autofocus to True before starting
        # This prefents reading False before the device replied with the first True
        device = self.__all_devices[serial_number]
        device.is_auto_focusing = True

        self.__send_ws_message(
            {
                "type": "AUTOFOCUS",
                "payload": {
                    "action": "START",
                    "serialNumber": serial_number,
                    "chamberType": chamber_type_number,
                },
            }
        )

        start_time = time.time()
        while device.is_auto_focusing:
            # Time out check
            current_waiting_time = time.time() - start_time
            if current_waiting_time > max_waiting_time and max_waiting_time > 0:
                raise TimeoutError(
                    f"""
                    Auto focusing took too long. It took {current_waiting_time}, max is {max_waiting_time}. 
                    """
                )

            # To not spam the service wait a bit
            time.sleep(0.1)

    def set_liveview(self, serial_number: str, state: bool = True) -> None:
        """
        Turn the liveview on or off

        serial_number: (str) the serial number of device you want to connect
        state: (bool) True = live view on
        """
        self.__send_ws_message(
            {
                "type": "LIVE_STREAM",
                "payload": {"serialNumber": serial_number, "enable": state},
            }
        )

    def open_liveview(self, serial_number: str) -> None:
        """
        Open the live view in the browser

        serial_number: (str) the serial number of device you want to connect
        """
        webbrowser.open(
            f"http://localhost:3333/cytosmartservice/live?serialNumber={serial_number}"
        )

    def set_zoom(self, serial_number: str, zoom_type: str = "IN") -> None:
        """
        Set zoom type by turning off or on binning.

        serial_number: (str) the serial number of device you want to connect
        zoom_type: (bool) str = IN or OUT
        """
        zoom_type = zoom_type.upper()
        assert zoom_type in [
            "IN",
            "OUT",
        ], f"zoom type needs to be IN or OUT not {zoom_type}"

        self.__send_ws_message(
            {
                "type": "ZOOM",
                "payload": {"serialNumber": serial_number, "action": zoom_type},
            }
        )

        # Toggle liveview to enforce the settings
        self.set_liveview(serial_number, False)
        self.set_liveview(serial_number, True)

    def set_focus(self, serial_number: str, focus_level: float = 0) -> None:
        """
        Set the relative z-position of the camera.
        And with that the focus.

        serial_number: (str) the serial number of device you want to connect
        focus_level: (float) between 0 and 1 where the camera need to be.
        """
        assert (
            focus_level <= 1 and focus_level >= 0
        ), f"Focus level needs to be between 0 and 1, not {focus_level}"

        current_channel = self.__all_devices[serial_number].active_channel
        self.__set_active_camera(serial_number, "BRIGHTFIELD")
        self.__send_ws_message(
            {
                "type": "FOCUS",
                "payload": {"serialNumber": serial_number, "value": focus_level},
            }
        )
        self.__set_active_camera(serial_number, current_channel)
        # Give device time to go to new focus
        if current_channel == "BRIGHTFIELD":
            time.sleep(0.5)
        else:
            time.sleep(2)

    def set_active_camera(
        self,
        serial_number: str,
        color_channel: str = "BRIGHTFIELD",
        max_waiting_time: float = 10,
    ) -> None:
        """
        Set a camera to active.
        You can switch between brightfield, red fluo, and green fluo
        This is only for fluo devices

        serial_number: (str) the serial number of device you want to connect
        focus_level: (float) between 0 and 1 where the camera needs to be.
        color_channel: (str) The camera you want to change the setting of.
            options: "BRIGHTFIELD", "RED", "GREEN"
            Default: "BRIGHTFIELD"
        max_waiting_time (float, optional): Maximum time it waits for the stage to arrive.
                If it is set to -1 it will never timeout
                Defaults to 10.
        """
        color_channel = color_channel.upper()
        assert color_channel in [
            "BRIGHTFIELD",
            "RED",
            "GREEN",
        ], f"color_channel needs to be BRIGHTFIELD, RED, or GREEN not {color_channel}"
        self.__set_active_camera(
            serial_number, color_channel, max_waiting_time=max_waiting_time
        )

    def __set_active_camera(
        self,
        serial_number: str,
        color_channel: str = "BRIGHTFIELD",
        max_waiting_time: float = 60,
    ) -> None:
        device = self.__all_devices[serial_number]
        if device.active_channel == color_channel:
            return
        self.__send_ws_message(
            {
                "type": "COLOR_CHANNEL",
                "payload": {
                    "serialNumber": serial_number,
                    "colorChannel": color_channel,
                },
            }
        )

        start_time = time.time()
        while True:
            if device.active_channel == color_channel:
                return
            time.sleep(0.1)
            current_waiting_time = time.time() - start_time
            if current_waiting_time > max_waiting_time and max_waiting_time > 0:
                raise TimeoutError(
                    f"""
                    Changing the channel took too long. It took {current_waiting_time}, max is {max_waiting_time}. 
                    """
                )

    def set_camera_settings(
        self,
        serial_number: str,
        color_channel: str = "BRIGHTFIELD",
        exposure: float = 10,
        gain: int = 20,
        brightness: int = 7500,
        focus_offset: float = 0.0,
    ) -> None:
        """
        Change the camera settings for 1 camera of 1 device.
        Brightfield devices have 1 camera; BRIGHTFIELD
        Fluo device have 3; BRIGHTFIELD, RED, and GREEN

        WARNING: This only changes settings of the camera,
            it does NOT activate it.
            Use set_active_camera to activate the camera you need.

        serial_number: (str) the serial number of device you want to connect
        exposure: (float) The time in milliseconds the camera is detecting light.
        gain: (int) The multiplication of the camera. (Fluo only)
            If very little light goes into the camera sensor make sure the gain is high.
        brightness: (int) Strength of the led when it is on (Fluo only)
        color_channel: (str) The camera you want to change the setting of.
            options: "BRIGHTFIELD", "RED", "GREEN"
            Default: "BRIGHTFIELD"
        focus_offset: (float) the difference in focus between brightfield and fluo.
            If focus is set to 0.4 and focus_offset for RED is set to 0.1 RED focus is 0.5 (Fluo only)
        """
        color_channel = color_channel.upper()
        assert color_channel in [
            "BRIGHTFIELD",
            "RED",
            "GREEN",
        ], f"color_channel needs to be BRIGHTFIELD, RED, or GREEN not {color_channel}"
        if color_channel == "BRIGHTFIELD":
            assert (
                0 < exposure and exposure <= 10
            ), f"exposure for BRIGHTFIELD needs to be between 0 and 10, not {exposure}"
        else:
            assert (
                0 < exposure and exposure <= 8000
            ), f"exposure for RED, or GREEN needs to be between 0 and 8000, not {exposure}"
        assert (
            0 < gain and gain < 100
        ), f"gain needs to be between 0 and 100, not {gain}"
        assert (
            0 < brightness and brightness <= 10000
        ), f"brightness needs to be between 0 and 10000, not {brightness}"

        # Turn on live stream and wait till it is one
        device = self.__all_devices[serial_number]

        self.set_liveview(serial_number, True)

        while device.live_stream == False:
            pass

        self.__send_ws_message(
            {
                "type": "CAMERA_SETTINGS",
                "payload": {
                    "serialNumber": serial_number,
                    "gain": gain,
                    "exposure": exposure,
                    "colorChannel": color_channel,
                    "brightness": brightness,
                    "focusOffset": focus_offset,
                },
            }
        )

    def set_flash_duration(self, serial_number: str, duration: int) -> None:
        """
        Setting the time in μs the led is on during the taking of a picture.

        Args:
        serial_number: (str) the serial number of device you want to connect
            duration (int): μs between 40 and 250
        """
        assert (
            40 <= duration and duration <= 250
        ), f"Duration needs to be between 40 and 250 μs, not {duration}"
        self.__send_ws_message(
            {
                "type": "OMNI_SET_FLASH_DURATION",
                "payload": {"serialNumber": serial_number, "duration": duration},
            }
        )

    def get_all_serial_numbers(self):
        """
        Returns all the serial numbers of the devices that are connected.
        """
        all_serial_numbers = []
        for serial_number in self.__all_devices.keys():
            device = self.__all_devices[serial_number]
            if device.is_connected:
                all_serial_numbers.append(serial_number)
        return all_serial_numbers

    def get_temperature(self, serial_number: str) -> float:
        """
        Returns the latest know temperature of the device.

        serial_number: (str) the serial number of device you want to connect
        """
        device = self.__all_devices[serial_number]
        return device.temperature

    def move_stage(
        self,
        serial_number: str,
        new_x: float,
        new_y: float,
        max_waiting_time: float = 60,
    ) -> None:
        """
        Moves the stage the the new position (Omni only)

        Args:
            serial_number: (str) the serial number of device you want to connect
            new_x (float): new x position in millimeters
            new_y (float): new y position in millimeters
            max_waiting_time (float, optional): Maximum time it waits for the stage to arrive.
                If it is set to -1 it will never timeout
                Defaults to 60.

        Raises:
            TimeoutError: Stage took longer then the max_waiting_time to go to new position
        """

        msg = {
            "type": "OMNI_MOVE_STAGE",
            "payload": {
                "serialNumber": serial_number,
                "X": float(new_x),
                "Y": float(new_y),
            },
        }
        self.__send_ws_message(msg)

        start_time = time.time()
        device = self.__all_devices[serial_number]
        while True:
            cur_x, cur_y = self.get_position(serial_number)
            is_moving = device.is_moving

            # If the device is not moving AND close enough stop
            if (
                abs(cur_x - new_x) < 0.1
                and abs(cur_y - new_y) < 0.1
                and is_moving == False
            ):
                break
            # If it is only not move but still not close send the message again
            elif is_moving == False:
                self.__send_ws_message(msg)

            # Time out check
            current_waiting_time = time.time() - start_time
            if current_waiting_time > max_waiting_time and max_waiting_time > 0:
                raise TimeoutError(
                    f"""
                    Moving stage took too long. It took {current_waiting_time}, max is {max_waiting_time}. 
                    Current state:
                    Position: {cur_x}, {cur_y}
                    Is_moving: {is_moving}
                    Wanted position: {new_x}, {new_y}
                    """
                )

            # To not spam the service wait a bit
            time.sleep(0.5)

    def get_position(self, serial_number: str) -> Tuple[float, float]:
        """
        Returns the latest known position of the device.

        serial_number: (str) the serial number of device you want to connect
        """

        self.__send_ws_message(
            {
                "type": "OMNI_REQUEST_STAGE_POSITION",
                "payload": {"serialNumber": serial_number},
            }
        )
        time.sleep(0.1)
        device = self.__all_devices[serial_number]
        if device.is_sleeping:
            return -1, -1
        return device.x, device.y

    def get_image(self, serial_number: str) -> Image.Image:
        """
        Get the current image of the camera.

        serial_number: (str) the serial number of device you want to connect
        """
        self.__activate(serial_number)
        self.set_liveview(serial_number, True)
        response = requests.get(
            f"http://localhost:3333/cytosmartservice/lastlive?serialNumber={serial_number}"
        )
        img = Image.open(BytesIO(response.content))

        return img

    def get_z_stack(
        self,
        serial_number: str,
        num_img: int = 10,
        start_focus: float = 0,
        stop_focus: float = 1,
    ) -> List[Image.Image]:
        """
        Creates a z-stack.
        It will take multiple different images on different focus levels.

        serial_number: (str) the serial number of device you want to connect
        num_img: (int) the amount of images in the z-stack
        start_focus: (float) The focus of the first image
        stop_focus: (float) the focus of the last image
        """

        result = []
        focus_step = (stop_focus - start_focus) / (num_img - 1)

        for focus_n in range(num_img):
            focus = start_focus + focus_n * focus_step
            focus = focus if focus <= 1 else 1

            print(f"focus level: {focus}")
            self.set_focus(serial_number, focus)
            img = self.get_image(serial_number)
            result.append(img)

        return result
